src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -250,28 +250,46 @@ private static boolean containsExcludedFlags(String testStr) {
        return false;
     }
 
+    // Return a temp file name if the archive file exists. The temp file operation may
+    // cause permission exception which prevents real archive being created.
+    private static String getArchiveFileName(String name) throws Exception {
+        String tempFileName = name + ".temp";
+        File tempFile = new File(tempFileName);
+        if (tempFile.exists()) {
+            tempFile.delete();
+        } else {
+            if (tempFile.createNewFile()) {
+                tempFile.delete();
+            }
+        }
+
+        File archiveFile = new File(name);
+        if (!archiveFile.exists()) {
+            return name;
+        }
+        return tempFileName;
+    }
+
     /**
     * called from jcmd VM.cds to dump static or dynamic shared archive
     * @param isStatic true for dump static archive or false for dynnamic archive.
     * @param fileName user input archive name, can be null.
     */
     private static void dumpSharedArchive(boolean isStatic, String fileName) throws Exception {
         String currentPid = String.valueOf(ProcessHandle.current().pid());
-        String archiveFile =  fileName != null ? fileName :
+        String archiveFileName =  fileName != null ? fileName :
             "java_pid" + currentPid + (isStatic ? "_static.jsa" : "_dynamic.jsa");
 
-        // delete if archive file aready exists
-        File fileArchive = new File(archiveFile);
-        if (fileArchive.exists()) {
-            fileArchive.delete();
-        }
+        String tempFileName = getArchiveFileName(archiveFileName);
+        File tempArchiveFile = new File(tempFileName);
```

Comment:
I think the logic is too complicated. We can always write the archive to a temp file, and then rename it to the actual file. Also, to be consistent with the other variables, it should be `tempArchiveFileName`

```
String tempArchiveFileName = archiveFileName + ".tmp";
```

Reply:
getArchiveFileName also operates on create/delete temp file, so if that fails, throw exception back to caller. Do you think the logic should be kept?

Reply:
How about: 
```
File tempArchiveFile = new File(tempArchiveFileName);
tempArchiveFile.createNewFile(); // will throw exception if we cannot write to the directory.
tempArchiveFile.delete();
```